runtime/lua/vim/lsp/util.lua

```diff
@@ -822,5 +839,52 @@ function M.character_offset(buf, row, col)
   return str_utfindex(line, col)
 end
 
+do
+  local all_choice_options = {}
+
+  local function save_choice_callback(unique_name, options, callback)
+    validate {
+      unique_name = { unique_name, 's' },
+      options = { options, 't' },
+      callback = { callback, 'f' }
+    }
+
+    all_choice_options[unique_name] = { callback = callback, options = options }
+  end
+
+  function M.choice_option_callback(unique_name)
+    local line = api.nvim_win_get_cursor(0)[1]
+    local option = all_choice_options[unique_name].options[line]
+    local callback = all_choice_options[unique_name].callback
+    api.nvim_win_close(0, false)
+    callback(option)
+  end
+
+  function M.choice_preview(unique_name, options, title_fn, callback)
+    validate {
+      unique_name = { unique_name, 's' },
+      options = { options, 't' },
+      title_fn = { title_fn, 'f' }
+    }
+    save_choice_callback(unique_name, options, callback)
+    local bufnr, winnr = M.focusable_preview(unique_name, function()
+      local titles = {}
+      for _, option in ipairs(options) do
+        table.insert(titles, title_fn(option))
+      end
+      return titles
+    end
+    )
+    vim.api.nvim_set_current_win(winnr)
+    vim.api.nvim_buf_set_keymap(
+      bufnr, 'n', '<CR>',
+      string.format('<cmd>lua vim.lsp.util.choice_option_callback("%s")<CR>', unique_name),
```

Comment:
```suggestion
      string.format('<cmd>lua vim.lsp.util.choice_option_callback(%q)<CR>', unique_name),
```

`%q` quotes the string correctly.